include/openssl/hpke.h

```diff
@@ -0,0 +1,499 @@
+# define OSSL_HPKE_MODE_PSK               1 /**< Pre-shared key mode */
+# define OSSL_HPKE_MODE_AUTH              2 /**< Authenticated mode */
+# define OSSL_HPKE_MODE_PSKAUTH           3 /**< PSK+authenticated mode */
+
+/*
+ * The (16bit) HPKE algorithn IDs
+ */
+# define OSSL_HPKE_KEM_ID_RESERVED         0x0000 /**< not used */
+# define OSSL_HPKE_KEM_ID_P256             0x0010 /**< NIST P-256 */
+# define OSSL_HPKE_KEM_ID_P384             0x0011 /**< NIST P-256 */
+# define OSSL_HPKE_KEM_ID_P521             0x0012 /**< NIST P-521 */
+# define OSSL_HPKE_KEM_ID_25519            0x0020 /**< Curve25519 */
+# define OSSL_HPKE_KEM_ID_448              0x0021 /**< Curve448 */
+
+# define OSSL_HPKE_KDF_ID_RESERVED         0x0000 /**< not used */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA256      0x0001 /**< HKDF-SHA256 */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA384      0x0002 /**< HKDF-SHA512 */
+# define OSSL_HPKE_KDF_ID_HKDF_SHA512      0x0003 /**< HKDF-SHA512 */
+# define OSSL_HPKE_KDF_ID_MAX              0x0003 /**< HKDF-SHA512 */
+
+# define OSSL_HPKE_AEAD_ID_RESERVED        0x0000 /**< not used */
+# define OSSL_HPKE_AEAD_ID_AES_GCM_128     0x0001 /**< AES-GCM-128 */
+# define OSSL_HPKE_AEAD_ID_AES_GCM_256     0x0002 /**< AES-GCM-256 */
+# define OSSL_HPKE_AEAD_ID_CHACHA_POLY1305 0x0003 /**< Chacha20-Poly1305 */
+# define OSSL_HPKE_AEAD_ID_MAX             0x0003 /**< Chacha20-Poly1305 */
+
+/* strings for modes */
+# define OSSL_HPKE_MODESTR_BASE       "base"    /**< base mode (1) */
+# define OSSL_HPKE_MODESTR_PSK        "psk"     /**< psk mode (2) */
+# define OSSL_HPKE_MODESTR_AUTH       "auth"    /**< sender-key pair auth (3) */
+# define OSSL_HPKE_MODESTR_PSKAUTH    "pskauth" /**< psk+sender-key pair (4) */
+
+/* strings for suite components - ideally these'd be defined elsewhere */
+# define OSSL_HPKE_KEMSTR_P256        "P-256"                /**< KEM id 0x10 */
+# define OSSL_HPKE_KEMSTR_P384        "P-384"                /**< KEM id 0x11 */
+# define OSSL_HPKE_KEMSTR_P521        "P-521"                /**< KEM id 0x12 */
+# define OSSL_HPKE_KEMSTR_X25519      SN_X25519              /**< KEM id 0x20 */
+# define OSSL_HPKE_KEMSTR_X448        SN_X448                /**< KEM id 0x21 */
+# define OSSL_HPKE_KDFSTR_256         "hkdf-sha256"          /**< KDF id 1 */
+# define OSSL_HPKE_KDFSTR_384         "hkdf-sha384"          /**< KDF id 2 */
+# define OSSL_HPKE_KDFSTR_512         "hkdf-sha512"          /**< KDF id 3 */
+# define OSSL_HPKE_AEADSTR_AES128GCM  LN_aes_128_gcm         /**< AEAD id 1 */
+# define OSSL_HPKE_AEADSTR_AES256GCM  LN_aes_256_gcm         /**< AEAD id 2 */
+# define OSSL_HPKE_AEADSTR_CP         LN_chacha20_poly1305   /**< AEAD id 3 */
+
+/**
+ * @brief ciphersuite combination
+ */
+typedef struct {
+    uint16_t    kem_id; /**< Key Encryption Method id */
+    uint16_t    kdf_id; /**< Key Derivation Function id */
+    uint16_t    aead_id; /**< AEAD alg id */
+} ossl_hpke_suite_st;
```

Comment:
Should still be:

``` C
typedef struct {
...
} OSSL_HPKE_SUITE;
```

Reply:
Fair 'nuff, will do that once I have the deterministic keygen stuff pushed up.

Reply:
that's done in https://github.com/openssl/openssl/pull/17172/commits/f016ba0584aed95c53fc4d8b30f5923d432aa321